teps/0040-capture-step-exit-code.md

```diff
@@ -0,0 +1,70 @@
+---
+status: proposed
+title: 'Capture Step Exit Code'
+creation-date: '2021-01-06'
+last-updated: '2021-01-06'
+authors:
+- '@pritidesai'
+- '@afrittoli'
+---
+
+# TEP-0040: Capture Step Exit Code
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+Tekton tasks are defined as a collection of steps in which each step can specify a container image to run.
+Steps are executed in order in which they are specified. One single step failure results in a task failure
+in turn results in a pipeline failure. Once a step results in failure, rest of the steps are not executed.
+
+Many common pipelines have requirement where a step failure must not halt the entire execution. In order to
+continue executing a task even if a step fail, pipeline authors have flexibility of wrapping an image and
+exiting that task with success. This changes the failing task into success and does not block further
+execution. This is a workaround and only works with images which can be wrapped. This workaround does not apply
+to off-the-shelf container images.
+
+As a pipeline execution engine, we want to support off-the-shelf container image as a step and provide
+an option to capture non-zero exit code. The pipeline author can choose to continue execution after capturing
+the non-zero exit code and make it available to access it after the execution.
+
+Issue: [tektoncd/pipeline#2800](https://github.com/tektoncd/pipeline/issues/2800)
+
+
+## Motivation
+
+It should be possible to easily use off-the-shelves (OTS) images as steps in Tekton tasks. A task author has no
+control on the image but may desire to capture an error and continue executing the pipeline.
```

Comment:
can you provide some examples of images that cannot be used today? or what it would look like to try to use them?

Reply:
Is the main issue with these off-the-shelf images that they sometimes don't have a shell? this was the motivation for TEP-0011 as well

(if the images have a shell, you can write your own script to do whatever you want with the exit code)

Reply:
Yes, indeed, the most difficult case is that of images that do not include a shell.
The wrapping script feels to me like a workaround though, I think it would be a nicer user experience if we allowed natively to trap the exit code.

Reply:
Container images are not necessarily written with a wrapping script in mind. 
If an image sets an entrypoint, it is usually the recommended way to use the image. To write a script that wraps the container, we need to inspect the container image, find out what the entry point is, and run the command in our script the same way the entry point would have, which is not something we should ask Task author to be doing. If the entrypoint changes, the task will have to be updated too.

We could have a mechanism that fetches the entrypoint from the docker image automatically and runs it... oh, wait, we already do, it's Tekton's entrypoint. This TEP proposes to add a flag in the entrypoint to capture the exit code - I don't think we should do this instead in a shell script that will be executed by entrypoint.

Reply:
kk - maybe we can discuss this more when we start looking at possible designs